### ISSUE TITLE
Can't create a new gem when using git on Windows

### ISSUE THREAD

**Issue by ronyman-com** [rubygems/rubygems#4567]
## Bundler settings

```
gem.test
  Set for the current user (C:/Users/Rony MAN/.bundle/config): "minitest"
gem.mit
  Set for the current user (C:/Users/Rony MAN/.bundle/config): true
gem.coc
  Set for the current user (C:/Users/Rony MAN/.bundle/config): true
gem.ci
  Set for the current user (C:/Users/Rony MAN/.bundle/config): false
gem.rubocop
  Set for the current user (C:/Users/Rony MAN/.bundle/config): "true"
gem.changelog
  Set for the current user (C:/Users/Rony MAN/.bundle/config): "true"
```

--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/rubygems/rubygems/search?q=No+such+file+or+directory+%40+rb_sysopen+-+C+%2FUsers%2FRony+MAN%2FOneDrive%2FRuby%2Fgetweb%2F.git%2FHEAD&type=Issues

If there aren't any reports for this error yet, please create copy and paste the report template above into a new issue. Don't forget to anonymize any private data! The new issue form is located at:
https://github.com/rubygems/rubygems/issues/new?labels=Bundler

**Comment by deivid-rodriguez** [comment]
Please copy the full output of the error.

**Comment by ronyman-com** [comment]
C:\Users\Rony MAN\OneDrive\Ruby>bundle gem getweb
Creating gem 'getweb'...
MIT License enabled in config
Code of conduct enabled in config
Changelog enabled in config
RuboCop enabled in config
Initializing git repo in C:/Users/Rony MAN/OneDrive/Ruby/getweb
usage: git init [-q | --quiet] [--bare] [--template=<template-directory>] [--shared[=<permissions>]] [<directory>]
--- ERROR REPORT TEMPLATE -------------------------------------------------------
# Error Report

## Questions

Please fill out answers to these questions, it'll help us figure out
why things are going wrong.

- **What did you do?**

  I ran the command `C:/Ruby30-x64/bin/bundle gem getweb`

- **What did you expect to happen?**

  I expected Bundler to...

- **What happened instead?**

  Instead, what happened was...

- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**

  I tried...

- **Have you read our issues document, https://github.com/rubygems/rubygems/blob/master/bundler/doc/contributing/ISSUES.md?**

  ...

## Backtrace

```
Errno::ENOENT: No such file or directory @ rb_sysopen - C:/Users/Rony MAN/OneDrive/Ruby/getweb/.git/HEAD
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/cli/gem.rb:196:in `read'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/cli/gem.rb:196:in `run'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/cli.rb:613:in `run'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/cli.rb:30:in `dispatch'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/cli.rb:24:in `start'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/exe/bundle:49:in `block in <top (required)>'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/lib/bundler/friendly_errors.rb:130:in `with_friendly_errors'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.16/exe/bundle:37:in `<top (required)>'
  C:/Ruby30-x64/bin/bundle:23:in `load'
  C:/Ruby30-x64/bin/bundle:23:in `<main>'
```

## Environment

```
Bundler       2.2.16
  Platforms   ruby, x64-mingw32
Ruby          3.0.1p64 (2021-04-05 revision 0fb782ee38ea37fd5fe8b1f775f8ad866a82a3f0) [x64-mingw32]
  Full Path   C:/Ruby30-x64/bin/ruby.exe
  Config Dir  C:/ProgramData
RubyGems      3.2.15
  Gem Home    C:/Ruby30-x64/lib/ruby/gems/3.0.0
  Gem Path    C:/Users/Rony MAN/.gem/ruby/3.0.0;C:/Ruby30-x64/lib/ruby/gems/3.0.0
  User Home   C:/Users/Rony MAN
  User Path   C:/Users/Rony MAN/.gem/ruby/3.0.0
  Bin Dir     C:/Ruby30-x64/bin
Tools
  Git         2.28.0.windows.1
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2021-04-09
Git SHA           3d7bfaff25
Released Version  true
```

## Bundler settings

```
gem.test
  Set for the current user (C:/Users/Rony MAN/.bundle/config): "minitest"
gem.mit
  Set for the current user (C:/Users/Rony MAN/.bundle/config): true
gem.coc
  Set for the current user (C:/Users/Rony MAN/.bundle/config): true
gem.ci
  Set for the current user (C:/Users/Rony MAN/.bundle/config): false
gem.rubocop
  Set for the current user (C:/Users/Rony MAN/.bundle/config): "true"
gem.changelog
  Set for the current user (C:/Users/Rony MAN/.bundle/config): "true"
```

--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/rubygems/rubygems/search?q=No+such+file+or+directory+%40+rb_sysopen+-+C+%2FUsers%2FRony+MAN%2FOneDrive%2FRuby%2Fgetweb%2F.git%2FHEAD&type=Issues

If there aren't any reports for this error yet, please create copy and paste the report template above into a new issue. Don't forget to anonymize any private data! The new issue form is located at:
https://github.com/rubygems/rubygems/issues/new?labels=Bundler

**Comment by deivid-rodriguez** [comment]
Apparently Windows `git` does not support passing a folder argument to it. This is quite surprising since it's a very old git feature (introduced in https://github.com/git/git/blob/311531c9de557d25ac087c1637818bd2aad6eb3a/Documentation/RelNotes/1.6.5.txt#L123-L124, 11 years ago).

To confirm this is the issue, can you install bundler-2.2.12 (the newest version before we started using `git init <target_folder>`) and try again using that? Also, can you manuall run `git init C:/Users/Rony MAN/OneDrive/Ruby/getweb` and verify whether you get any errors?

**Comment by ronyman-com** [comment]
I did follow your instruction. But nothing worth...

**Comment by deivid-rodriguez** [comment]
What do you mean? Could you explain the results you got? Thanks.

**Comment by deivid-rodriguez** [comment]
This same error has been reported at #4622, and Windows is not being used there, so I'll close this one in favour of the new issue.

**Comment by Secreto31126** [comment]
@deivid-rodriguez I know you closed this particular issue, but I just tried doing this...

> Also, can you manuall run `git init C:/Users/Rony MAN/OneDrive/Ruby/getweb` and verify whether you get any errors?

...on my computer, and the command failed in a directory _with a whitespace and full path_. Then I tried to run the same command on another folder _without whitespaces and full path_ and it runs successfully. I also tried to run _without the full path and with whitespaces_, and again it worked correctly.

Obviously I retried running `bundle gem` on the new directory _without whitespaces_ and it worked as expected. Just in case I retried `bundle gem` on the one _with whitespace_ and it failed again.

I tried using git on WSL, and it worked perfectly _with the whitespaces and full path_, so I don't know if it has any relation with the other issue you mentioned.

Last, I want to mention something I noticed from the output of git:

`
PS C:\Users\myname\Documents\Ruby> git init 'C:\Users\myname\Documents\Ruby\New Test\'
fatal: cannot mkdir C:\Users\myname\Documents\Ruby\New Test": Invalid argument
`

Notice that extra `"` at the end of the path? I might be overlooking at it, but I have the feeling that it shouldn't be there.

Who knows? Maybe we found a glitch in git :)

`git version 2.24.1.windows.2` and `git version 2.32.0.windows.2`, just updated to the latest version and it still has the same issue.

**Comment by Secreto31126** [comment]
Ohh, I just noticed that I was running the command with \ at the end of the command, so the " is being escaped. I tried git with the full path with whitespaces and _without_ the last \, and now it worked as expected.

Maybe there's the issue? Now I'm thoughtless.

**Comment by deivid-rodriguez** [comment]
Thanks for your investigation @Secreto31126! That's useful.

It seems the same as https://github.com/git-for-windows/git/issues/1444? I guess we can workaround this by removing the trailing slash if it's there because shelling out to `git`.

Can you copy the exact `bundle gem` command that you're running, and the output you get?

**Comment by Secreto31126** [comment]
`PS C:\Users\myname\Documents\Ruby\White Space> bundle gem test`

Creating gem 'test'...
MIT License enabled in config
Code of conduct enabled in config
Changelog enabled in config
RuboCop enabled in config
Initializing git repo in C:/Users/myname/Documents/Ruby/White Space/test
usage: git init [-q | --quiet] [--bare] [--template=<template-directory>] [--shared[=<permissions>]] [<directory>]
--- ERROR REPORT TEMPLATE -------------------------------------------------------
# Error Report

## Questions

Please fill out answers to these questions, it'll help us figure out
why things are going wrong.

- **What did you do?**

  I ran the command `C:/Ruby30-x64/bin/bundle.cmd gem test`

- **What did you expect to happen?**

  I expected Bundler to create the template

- **What happened instead?**

  Instead, what happened was this error

- **Have you tried any solutions posted on similar issues in our issue tracker, stack overflow, or google?**

  I tried helping in a Github issue

- **Have you read our issues document, https://github.com/rubygems/rubygems/blob/master/bundler/doc/contributing/ISSUES.md?**

  Too much work :)

## Backtrace

```
Errno::ENOENT: No such file or directory @ rb_sysopen - C:/Users/myname/Documents/Ruby/White Space/test/.git/HEAD
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/cli/gem.rb:196:in `read'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/cli/gem.rb:196:in `run'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/cli.rb:593:in `run'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/cli.rb:30:in `dispatch'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/cli.rb:24:in `start'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.22/libexec/bundle:49:in `block in <top (required)>'
  C:/Ruby30-x64/lib/ruby/3.0.0/bundler/friendly_errors.rb:128:in `with_friendly_errors'
  C:/Ruby30-x64/lib/ruby/gems/3.0.0/gems/bundler-2.2.22/libexec/bundle:37:in `<top (required)>'
  C:/Ruby30-x64/bin/bundle.cmd:31:in `load'
  C:/Ruby30-x64/bin/bundle.cmd:31:in `<main>'
```

## Environment

```
Bundler       2.2.22
  Platforms   ruby, x64-mingw32
Ruby          3.0.2p107 (2021-07-07 revision 0db68f023372b634603c74fca94588b457be084c) [x64-mingw32]
  Full Path   C:/Ruby30-x64/bin/ruby.exe
  Config Dir  C:/ProgramData
RubyGems      3.2.22
  Gem Home    C:/Ruby30-x64/lib/ruby/gems/3.0.0
  Gem Path    C:/Users/myname/.local/share/gem/ruby/3.0.0;C:/Ruby30-x64/lib/ruby/gems/3.0.0
  User Home   C:/Users/myname
  User Path   C:/Users/myname/.local/share/gem/ruby/3.0.0
  Bin Dir     C:/Ruby30-x64/bin
Tools
  Git         2.32.0.windows.2
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2021-08-04
Git SHA           unknown
Released Version  false
```

## Bundler settings

```
gem.changelog
  Set for the current user (C:/Users/myname/.bundle/config): true
gem.ci
  Set for the current user (C:/Users/myname/.bundle/config): "github"
gem.coc
  Set for the current user (C:/Users/myname/.bundle/config): true
gem.mit
  Set for the current user (C:/Users/myname/.bundle/config): true
gem.rubocop
  Set for the current user (C:/Users/myname/.bundle/config): "true"
gem.test
  Set for the current user (C:/Users/myname/.bundle/config): false
```

**Comment by deivid-rodriguez** [comment]
Alright, I can repro this now on Linux, so it doesn't seem Windows specific in the end. Closing again in favour of #4622, will fix shortly.

**Comment by deivid-rodriguez** [comment]
#4816 should fix this!

**Comment by deivid-rodriguez** [comment]
Thanks so much for the help with the reproduction! :heart:

**Comment by Secreto31126** [comment]
I'm happy to help 😄
